mgr/dashboard: Add SSO through oauth2 protocol

Pegonzal opened the pull request:
This PR intends to enable SSO in the dashboard through oauth2 protocol thanks to the `mgmt-gateway` and `oauth2-proxy` service. You can configure them following the documented steps in #58460

To access the dashboard with oauth2 SSO, you'll have to enable it:
 `ceph dashboard sso enable oauth2`

From the IDP configured of your choice:

- You'll need to set the Valid reditrect URI: `https://<host_name>|<IP_address>/oauth2/callback` wich has to be the same configured on the `oauth2-proxy` service, for the `redirec urls` field

- To login you need a valid user role, for administrator role you will have to configure the IDP's user with an 'administrator' role or 'read-only' for a read-only access

Possible improvements **TODO**: 

- [ ] Pluggable backend: [proposal](https://github.com/ceph/ceph/pull/58456#pullrequestreview-2230565811)
- [ ] Refactor auth + authen to [remove oauth2 user create/delete pattern](https://github.com/ceph/ceph/pull/58456#discussion_r1732583607)
- [ ] Mantain standalond backend: [Remove frontend redirect](https://github.com/ceph/ceph/pull/58456#discussion_r1732609434)
- [ ] Base auth router: [propsal](https://github.com/ceph/ceph/pull/58456#discussion_r1742005499)
- [ ] Refactor the JwtManager and OAuth2: [base class](https://github.com/ceph/ceph/pull/58456#discussion_r1750729272) and [factory pattern](https://github.com/ceph/ceph/pull/58456#discussion_r1750646987)
- [ ] Decorator for loggin on function start and stop: [proposal](https://github.com/ceph/ceph/pull/58456#discussion_r1751530433)
- [ ] Improve Idp's user roles mappging to Dashboard's user roles
- [ ] Redirect to Idp's login on token expiration


-------------

 
[screen-capture (20).webm](https://github.com/user-attachments/assets/431a1a44-f4c6-49ac-bf60-b755d1dfd331)


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

Pegonzal committed:
mgr/dashboard: add SSO through oauth2 protocol

Fixes: https://tracker.ceph.com/issues/66900
Signed-off-by: Pedro Gonzalez Gomez <pegonzal@redhat.com>

nizamial09 commented:
@epuertat can you give a review to this PR when you get a chance, thanks?

epuertat reviewed (commented):
@Pegonzal sorry for the late review! Believe or not, my last 30 mins before holidays is the first time I can find time to look at this 😅 

If it works, from a functional PoV, I have nothing else to add, so LGTM! 👍🏼 

If you ask me about the implementation approach, I'd mention the following:

For backward-compatibility/rollback scenarios/situations in which a new feature doesn't work, I'd clearly isolate the new functionality from the existing one. Given that authentication is a critical function in a system, issues at that stage might lead to inability to log in, or security breaches (I'm not saying that this code contains those... but I'm not saying otherwise either).

How would I approach this? I'd isolate the new functionality from the existing one. From my understanding the new SSO auth scheme overlaps with the local one (we even have a SAML2 SSO, which we could perhaps deprecate after merging this, CC: @nizamial09 ?).

And how would I specifically address this? Looking at how others do that. Usually authentication is considered, in many systems as a pluggable back-end (e.g.: [PAM](https://en.wikipedia.org/wiki/Linux_PAM) in Linux, [mod_authn_*](https://httpd.apache.org/docs/2.4/mod/#A) in Apache, ...).

One possible design here would be to identify the interface between the core Dashboard and the authentication back-end.

The Dashboard just needs to read a config option where the type of auth backend is defined and configured (e.g.: `mgr/dashboard/authentication_backend = local`, `mgr/dashboard/authentication_backend = saml2:<csv_settings>`, `mgr/dashboard/authentication_backend = oauth2:<csv_settings>`).

Then when starting, it may try to dynamically import a Python file named `.../mgr/dashboard/auth/local.py`, `.../mgr/dashboard/auth/saml2.py` or `.../mgr/dashboard/auth/oauth2.py`.

Each of these files contains a single class that implements an abstract class. The challenge here is to identify the minimum number of abstract methods that are required to isolate the different back-ends (local, saml2, oauth2).

```py
class AbstractAuthenticationProvider(metaclass=ABCMeta): 	
  @abstracmethod
  def authenticate_user(username, password) -> Token:
    pass
```

The benefit of this approach goes beyond what I mentioned initially. It also helps to implement integration tests, since you don't have to test the Dashboard code, but all the methods that each back-end provides.

Hope it helps!

nizamial09 commented:
> From my understanding the new SSO auth scheme overlaps with the local one (we even have a SAML2 SSO, which we could perhaps deprecate after merging this, CC: @nizamial09 ?).

Yup, I think moving forward we can try and deprecate it, but i don't know if it should be right after? Probably we'll need to issue a deprecating notice and eventually deprecate it since users might be relying on it.

> The benefit of this approach goes beyond what I mentioned initially. It also helps to implement integration tests, since you don't have to test the Dashboard code, but all the methods that each back-end provides.

@Pegonzal I agree with @epuertat's comments. We definitely need to isolate the behavior so that even if the new change breaks, we can simply disable that so the older ability will be restored. But given the time constraints I am open to merge this here in the main branch and open trackers for the improvements mentioned here but it'd be better if there are tests written for the older and newer behavior so we know atleast everything is working functionally.

nizamial09 commented:
also can you guys post a recording of the manual testing?

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -160,13 +165,38 @@ def logout(self):
         }
 
+    def _oauth2_logout(self):
+        idp_domain = "192.168.100.100:8080"
```
Yes, I'll be fixing that. My last commit was mainly to let Afreen test it, but introduced some hardcoded values.

Will update it.

Pegonzal commented:
> also can you guys post a recording of the manual testing?

There is a video on the PR description, I'll update it with a more recent one, or do you mean something else?

nizamial09 commented:
> > also can you guys post a recording of the manual testing?
> 
> There is a video on the PR description, I'll update it with a more recent one, or do you mean something else?

i completely missed it, sorry. that should be fine.

rkachach commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -186,7 +224,39 @@ def check(self, token):
-            'login_url': self._get_login_url(),
-            'cluster_status': ClusterModel.from_db().dict()['status']
-        }
+        elif hasattr(cherrypy.request, 'cookie') and '_oauth2_proxy' in cherrypy.request.cookie and \
+          cherrypy.request.cookie['_oauth2_proxy'].value:
+            orch = OrchClient()
+            if not orch.services.list_daemons(daemon_type='mgmt-gateway'):
+                logger.info('mgmt-gateway is not configured, can not access Dashboard with oauth2 SSO')
+                return default_return()
+            access_token = cherrypy.request.headers.get('X-Access-Token')
+            if not access_token:
+                return default_return()
+            JwtManager.oauth2_token = True
+            token_payload = json.loads(base64.urlsafe_b64decode(access_token.split(".")[1] + "===="))
+            if not 'sub' in token_payload:
+                return default_return()
+            try:
+                user = AuthManager.get_user(token_payload['sub'])
+            except UserDoesNotExist:
+                user = mgr.ACCESS_CTRL_DB.create_user(token_payload['sub'], None, token_payload['name'], token_payload['email'])
+                if 'resource_access' in token_payload:
+                    # Find the first value where the key is not 'account'
+                    user_roles = next(value for key, value in token_payload['resource_access'].items() if key != "account")
+                    if user_roles:
+                        user.set_roles([Role.map_to_system_roles(user_roles['roles'][0])])
+                else: #if no valid role is provided set readonly as default
+                    user.set_roles([READ_ONLY_ROLE])
+                user.last_update=token_payload['iat'] #set user last update to token time issued
+                mgr.ACCESS_CTRL_DB.save()
+            if user:
+                url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
+                self._set_token_cookie(url_prefix, access_token)
+                return {
+                    'username': user.username,
+                    'permissions': user.permissions_dict(),
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
+                    'pwdUpdateRequired': user.pwd_update_required
+                }
```
this function clearly handle the case of `oauth2-proxy`, I'd recommend having something like (please, read the code and modify it carefully .. I din't test it, just a re-factorization suggestion):

```
    def parse_user_roles(token_payload):
        # first we cehck for Client Roles section
        if 'resource_access' in token_payload:
            # Find the first value where the key is not 'account'
            user_roles = next(value for key, value in token_payload['resource_access'].items() if key != "account")
            return [Role.map_to_system_roles(user_roles['roles'][0])] if ('roles' in user_roles and user_roles['roles']) else []
        # if it doesn't exist we cehck for Realm Roles section
         elif 'resource_access':
             return ..

     def get_user(token_payload):
        try:
            user = AuthManager.get_user(token_payload['sub'])
        except UserDoesNotExist:
            user = mgr.ACCESS_CTRL_DB.create_user(token_payload['sub'], None, token_payload['name'], token_payload['email'])
            user_roles = parse_user_roles(token_payload)
            if user_roles:
                user.set_roles(user_roles)
            else:
                user.set_roles([READ_ONLY_ROLE])
            user.last_update=token_payload['iat'] #set user last update to token time issued
            mgr.ACCESS_CTRL_DB.save()
        return user

    @RESTController.Collection('POST', query_params=['token'])
    @EndpointDoc("Check token Authentication",
                 parameters={'token': (str, 'Authentication Token')},
                 responses={201: AUTH_CHECK_SCHEMA})
    def check(self, token):
        JwtManager.oauth2_token = False

        def default_return():
           return {
                    'login_url': self._get_login_url(),
                    'cluster_status': ClusterModel.from_db().dict()['status']
                }

        if token:
            user = JwtManager.get_user(token)
            if user:
                return {
                    'username': user.username,
                    'permissions': user.permissions_dict(),
                    'sso': mgr.SSO_DB.protocol == 'saml2',
                    'pwdUpdateRequired': user.pwd_update_required
                }
        elif is_oidc_request(cherrypy.request):

            orch = OrchClient()
            if not orch.services.list_daemons(daemon_type='mgmt-gateway'):
                logger.info('mgmt-gateway is not configured, can not access Dashboard with oauth2 SSO')
                return default_return()

            access_token = cherrypy.request.headers.get('X-Access-Token')
            if not access_token:
                return default_return()
            JwtManager.oauth2_token = True
            token_payload = json.loads(base64.urlsafe_b64decode(access_token.split(".")[1] + "===="))
            if not 'sub' in token_payload:
                return default_return()

            user = get_user()
            url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
            self._set_token_cookie(url_prefix, access_token)
            return {
                'username': user.username,
                'permissions': user.permissions_dict(),
                'sso': mgr.SSO_DB.protocol == 'saml2',
                'pwdUpdateRequired': user.pwd_update_required
            }

        return default_return()

```
I would also look into adding the part: 

```
        if token:
            user = JwtManager.get_user(token)
            if user:
                return {
                    'username': user.username,
                    'permissions': user.permissions_dict(),
                    'sso': mgr.SSO_DB.protocol == 'saml2',
                    'pwdUpdateRequired': user.pwd_update_required
                }
```
into the get user ... as it has some which being repeated later .. but that's more risky in as modification because it changes the behaviour and the lines:

``` 
           url_prefix = 'https' if mgr.get_localized_module_option('ssl') else 'http'
            self._set_token_cookie(url_prefix, access_token)
``` 

are only called for oauth2-proxy case...

Pegonzal replied:
Thanks Redo, guided by your suggestions I refactored the code

afreen23 commented:
Hi I tested this PR, noticed following things:

* After login and making the keyclock user as admin, cant login to actual dashboard not behin nginx as "admin" user.
* The login page started flickering for me when logged out at times mostly when logged out after first login

Thanks

Pegonzal commented:
Thanks for testing @afreen23 ! Following on your comments:

> * After login and making the keyclock user as admin, cant login to actual dashboard not behin nginx as "admin" user.

Could you elaborate on this? What do you mean with 'actual dashboard'?

> * The login page started flickering for me when logged out at times mostly when logged out after first login

Looking into it!

afreen23 commented:
> Thanks for testing @afreen23 ! Following on your comments:
> 
> > * After login and making the keyclock user as admin, cant login to actual dashboard not behin nginx as "admin" user.
> 
> Could you elaborate on this? What do you mean with 'actual dashboard'?

The one not running behind nginx. I am assuming we are making dashboard available at both endpoints.
Correct me if I am wrong,
> 
> > * The login page started flickering for me when logged out at times mostly when logged out after first login
> 
> Looking into it!

epuertat commented on src/pybind/mgr/dashboard/services/access_control.py:
```diff
@@ -193,6 +194,15 @@ def from_dict(cls, r_dict):
+    def map_to_system_roles(cls, role):
+        matches = []
+        for rn in SYSTEM_ROLES_NAMES:
+            closest_match = difflib.get_close_matches(role.lower(), SYSTEM_ROLES_NAMES[rn], n=1, cutoff=0.8)
```
That's a clear no. In [Security](https://cheatsheetseries.owasp.org/cheatsheets/Secure_Product_Design_Cheat_Sheet.html), you don't play with fuzzy matches. Either authorization scopes match 100% or they don't.

What's the sense of allowing `damin` role to be granted `admin` rights?

Pegonzal replied:
That make sense.
 The `get_close_matches` cutoff parameter was set to `0.8` (pretty strict) to forbid things like `damin` and allow `admin` even though the real role name is `administator`

But the security risk is clear. I have removed it

epuertat commented on src/pybind/mgr/dashboard/services/access_control.py:
```diff
@@ -283,6 +293,13 @@ def from_dict(cls, r_dict):
     GANESHA_MGR_ROLE.name: GANESHA_MGR_ROLE,
 }
 
+# static name-like roles list for role mapping
+SYSTEM_ROLES_NAMES = {
+    ADMIN_ROLE: [ADMIN_ROLE.name, 'admin'],
+    READ_ONLY_ROLE: [READ_ONLY_ROLE.name, 'read', 'guest', 'monitor']
+}
```
This is ok. If you want to make it more flexible, you can check [how Grafana does OAuth2 role mapping](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#configure-role-mapping). We have support in CentOS 9 for a [Python JMESPath parser](https://pypi.org/project/jmespath/).

Pegonzal replied:
For now I have removed this, we'll look into itroducing something similar like Grafana in the future

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -30,6 +32,7 @@ class JwtManager(object):
     JWT_TOKEN_TTL = 28800  # default 8 hours
     JWT_ALGORITHM = 'HS256'
     _secret = None
+    oauth2_token = False
```
Is this a per-user or a server global setting? If you don't know why I'm asking this look closely at the line underneath.

Pegonzal replied:
It was a global setting. I have removed it and using the mgr.SSO_DB now

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
             raise cherrypy.HTTPError(403, "You don't have permissions to "
                                           "access that resource")
+
+class Oauth2JWTManager(object):
```
Redundant since 2.7:
```suggestion
class Oauth2JWTManager:
```

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
+    @classmethod
+    def _create_user(cls):
+        user_roles = cls.get_user_roles()
+        user = mgr.ACCESS_CTRL_DB.create_user(cls._token_payload['sub'], None, cls._token_payload['name'], cls._token_payload['email'])
+        user.set_roles([Role.map_to_system_roles(user_roles[0])])
+
+        # set user last update to token time issued
+        user.last_update = cls._token_payload['iat']
+        mgr.ACCESS_CTRL_DB.save()
```
Why? Why is OAuth2 SSO mingling with the local user DB? The purpose of SSO is to fully offload the user management to a external entity, not to juggle with all the things together. All the user metadata should be kept in some client-side session token (JWT, whatever, ...).

Pegonzal replied:
The idea with the introduction of this new SSO mechanism was to modify as little as possible the current login, authorization and SSO (SAML2).

The current code uses the JWT token to get the user info and then makes use of the user to check for permission and for it's authorization.
I ended up with this wierd create/delete user, that would be used to authorize in the dashboard.

For now I'm not really sure if I should keep it

epuertat replied:
While drafting the design for the long-term JWT TTLs, I found that the implementation of JWT in the Dashboard is quite peculiar/non-standard (mixing client and server side state). Technically, once the JWT has been granted to the client, the server shouldn't perform any credential (claim) revalidation until JWT expiration.

That said, for OAuth2 we should at least avoid adding/deleting users to the mon DB. That's extra load for the mon DB that was not present with the local auth.

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
+        # check for client roles
+        if 'resource_access' in cls._token_payload:
+            # Find the first value where the key is not 'account'
+            user_roles = next(value['roles'] for key, value in cls._token_payload['resource_access'].items() if key != "account")
```
This might trigger a `StopIteration` exception unless you provide a default value: `next(iterator, default)`.

Pegonzal replied:
Added `user_roles` as default value

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
+        if token_payload:
+            cls._token_payload = token_payload
+        elif cls._token:
+            cls._token_payload = json.loads(base64.urlsafe_b64decode(cls._token.split(".")[1] + "===="))
```
This appears 4 times in this PR, why not providing a helper method to avoid repetition and increase consistency and readability:
```suggestion
            cls._token_payload = cls.get_token_payload(cls._token.split(".")[1])
```

Pegonzal replied:
Added helper function

rkachach commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -186,7 +227,22 @@ def check(self, token):
-            'login_url': self._get_login_url(),
-            'cluster_status': ClusterModel.from_db().dict()['status']
-        }
+        elif hasattr(cherrypy.request, 'cookie') and '_oauth2_proxy' in cherrypy.request.cookie and \
+          cherrypy.request.cookie['_oauth2_proxy'].value:
```
Please, remember to refactor this check on a function such as `is_oauth2_request(cherrypy.request)`

Pegonzal replied:
refactored it

rkachach commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
                                           "access that resource")
+
+class Oauth2JWTManager(object):
+    _token = ''
+    _token_payload = {}
```
Do we really need this attributes to be at class level? isn't it enough to just have them as object level?

Pegonzal replied:
Changed it to be at object level

rkachach commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
+    @classmethod
+    def _create_user(cls):
+        user_roles = cls.get_user_roles()
+        user = mgr.ACCESS_CTRL_DB.create_user(cls._token_payload['sub'], None, cls._token_payload['name'], cls._token_payload['email'])
```
this class needs mgr but we are not passing it as an argument! .. please, consider having a normal object instead of using the static `@classmethod` to call all the methods unless really needed.

Pegonzal replied:
Changed it

rkachach commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +288,64 @@ def _check_authorization(self, username):
+    def set_token(cls, token):
+        cls._token = token
+        cls.set_token_payload()
+        JwtManager.oauth2_token = True
```
sounds like this is repeated: you area already setting this variable inside the method `set_token_payload` ... and logic is different because inside the method you are setting it conditionally!

Pegonzal replied:
updated it

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -186,7 +227,22 @@ def check(self, token):
-            'login_url': self._get_login_url(),
-            'cluster_status': ClusterModel.from_db().dict()['status']
-        }
+        elif hasattr(cherrypy.request, 'cookie') and '_oauth2_proxy' in cherrypy.request.cookie and \
```
Added a new command to enable and configure oauth2

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -186,7 +227,22 @@ def check(self, token):
+                return {
+                    'username': user.username,
+                    'permissions': user.permissions_dict(),
+                    'sso': mgr.SSO_DB.protocol == 'saml2',
```
Kept the same fucntionality and return paths

epuertat commented on src/pybind/mgr/dashboard/services/oauth2_sso.py:
```diff
@@ -0,0 +1,71 @@
+    if cmd['prefix'] == 'dashboard sso disable':
+        mgr.SSO_DB.protocol = ''
+        mgr.SSO_DB.save()
+        return 0, 'SSO is "disabled".', ''
```
```suggestion
        return HandleCommandResult(stdout='SSO is "disabled"')
```

epuertat commented on src/pybind/mgr/dashboard/services/oauth2_sso.py:
```diff
@@ -0,0 +1,71 @@
+            mgr.SSO_DB.protocol = 'oauth2'
+            mgr.SSO_DB.save()
+            return 0, 'SSO is "enabled" with "oauth2" protocol.', ''
+        return -errno.EPERM, '', 'Single Sign-On is not configured: ' \
+            'use `ceph dashboard sso setup oauth2`'
```
Backtick "`" is great for Markdown formatting, but in CLI output it's probably better to just use ' and ".
```suggestion
        return HandleCommandResult(
            -errno.EPERM,
            stderr="Single Sign-On not configured: use 'ceph dashboard sso setup oauth2'")
```

epuertat commented on src/pybind/mgr/dashboard/services/oauth2_sso.py:
```diff
@@ -0,0 +1,71 @@
+        return 0, 'SSO is "disabled".', ''
+
+    if cmd['prefix'] == 'dashboard sso enable oauth2':
+        configured = _is_sso_configured()
+        if configured:
```
Python is all about clarity, but also conciseness. Unless you have to reuse the `configured` var later, this is better:
```suggestion
        if _is_sso_configured():
```

epuertat commented on src/pybind/mgr/dashboard/services/oauth2_sso.py:
```diff
@@ -0,0 +1,71 @@
+from .. import mgr
+
+
+class Oauth2:
+    def __init__(self, realm_name=None):
+        self.realm_name = realm_name
+
+    def to_dict(self):
+        return {
+            'realm_name': self.realm_name
+    }
```
As we no longer have to support Python 3.6, >=3.7 brings `dataclasses`:
```suggestion
from dataclasses import dataclass, asdict

@dataclass
class Oauth2:
    realm_name: str
```
If you don't want to use `dataclasses.asdict()` you can hard-wire that as:
```suggestion
from dataclasses import dataclass, asdict

@dataclass
class Oauth2:
    realm_name: str
    to_dict = asdict
```

epuertat commented on src/pybind/mgr/dashboard/services/sso.py:
```diff
@@ -46,17 +48,19 @@ class SsoDB(object):
     SSODB_CONFIG_KEY = "ssodb_v"
 
-    def __init__(self, version, protocol, saml2):
+    def __init__(self, version, protocol, saml2, oauth2):
```
As it doesn't make sense (and it'll probably not work) to have both SAML2 and OAuth2 together:
```suggestion
    def __init__(self, version, protocol: Literal["saml2", "oauth2"], config: Union[SAML2Config, OAuth2Config]):
```

epuertat commented on src/pybind/mgr/dashboard/services/sso.py:
```diff
@@ -46,17 +48,19 @@ class SsoDB(object):
             db = {
                 'protocol': self.protocol,
                 'saml2': self.saml2.to_dict(),
+                'oauth2': self.oauth2.to_dict(),
```
Internally you could still save OAuth2 config to a new entry, like this, in order to make this DB schema backward-compatible.

epuertat reviewed (changes requested):
Good work. This one looks much better! There are still things that need to be fixed.

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,71 @@
+        if self._meets_login_requirements():
+            access_token = cherrypy.request.headers.get('X-Access-Token')
+            if not access_token:
+                return 'Failed to login: Missing access token'
```
You can raise exceptions (HTTP errors) with custom text messages. The error handling stack is rather complicated (what a novelty, huh?), but eventually all Dashboard exceptions end up being caught by an exception handler:

https://github.com/ceph/ceph/blob/f689bb8f7f65df5848006bdee25da05ccb33d769/src/pybind/mgr/dashboard/services/exception.py#L44-L63

If you want to add custom text to the response body, here you have the [Cherrypy doc](https://docs.cherrypy.dev/en/latest/pkg/cherrypy._cperror.html).

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,71 @@
+        except (AttributeError, KeyError):
+            return False
+
+        orch = OrchClient().instance()
+        if not orch.services.list_daemons(daemon_type='oauth2-proxy') or not orch.services.list_daemons(daemon_type='mgmt-gateway'):
+            logger.warning('Can not access Dashboard with oauth2 SSO, mgmt-gateway and oauth2-proxy services are required')
+            return False
+        return True
```
I'd say yes, let's remove that part. From the Dashboard side, it only needs a new HTTP header, right? Both approaches are valid (fail-fast vs. fail-safe systems), but in this case it creates a dependency that it's not real (no Dashboard SSO code invoked the orchestrator).

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
+            cls._create_user()
+        return cls._user
+
+    @classmethod
+    def _create_user(cls):
+        if cls._token_payload:
+            try:
+                user = mgr.ACCESS_CTRL_DB.create_user(cls._token_payload['sub'], None, cls._token_payload['name'], cls._token_payload['email'])
+            except UserAlreadyExists:
+                user = mgr.ACCESS_CTRL_DB.get_user(cls._token_payload['sub'])
+            user.set_roles(cls.get_user_roles())
+            # set user last update to token time issued
+            user.last_update = cls._token_payload['iat']
+            cls._user = user
```
I hate this, but I understand it's needed because the local JWT/Auth manager didn't set/read the roles from the JWT payload...

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
+class Oauth2JWTManager:
+    _token = None
+    _token_payload = None
+    _user = None
```
I'd bet you to test this with multiple Dashboard users logged in... Sooner or later 2 threads will concurrently overwrite this and you'll get credentials from 1 user accessed by the other one.

That's why the JWTManager had (which was a clumsy thing, since `cherrypy.request` is already a thread-local object):
```py
    LOCAL_USER = threading.local()
```

Pegonzal replied:
Using `cherrypy.request` now

epuertat commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
+    _token_payload = None
+    _user = None
+
+    @classmethod
+    def initialize_from_token(cls, token):
+        if not JwtManager.is_blocklisted(token):
+            cls._token = token
+            cls._token_payload = cls.get_token_payload()
+            cls._user = cls.get_user()
```
When you create a class with only `@classmethod`s and an `init` method that initializes the class variables... Then that's a clear code smell. Objects are usually local, classes are global. Modifying a global class in a multi-threaded context and hoping that it'll behave as a regular object, that's a recipe for disaster.

We either use local objects, or thread-local global objects (like `cherrypy.request`) to store/fetch stuff.

Classes are meant to be instatiated (the JWTManager is an infamous examples and hopefully we'll get rid of it in the near future).

Pegonzal replied:
Using cherrypy.request now

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+    def login(self):
+        if self._meets_login_requirements():
+            access_token = cherrypy.request.headers.get('X-Access-Token')
+            if not access_token:
+                logger.error('Failed to login: Missing access token')
+                raise cherrypy.HTTPError(400, 'Failed to login: Missing access token')
```
> Even more, the dashboard_exception_handler() print logs for most exceptions,

Looks like it only raises the given exception?
https://github.com/ceph/ceph/blob/bba65b233b90d20d9ab694d2201126759501459c/src/pybind/mgr/dashboard/services/exception.py#L49-L51

Or am i not looking it right?

epuertat replied:
Yes, but there's this other Cherrypy Tool that takes care of printing logs:

https://github.com/ceph/ceph/blob/c9bd62a24819d4f5af62e610fe419a55be110b1f/src/pybind/mgr/dashboard/tools.py#L29-L122

epuertat replied:
The above means that the 400 will be logged as an event log (this usual structured, whitespace separated, log line: `[dashboard INFO request] [<ip>:<port>] [<method>] [<http_status>] [<latency>] [<size>] [<request_id>] <path>`), but it'll also print the text through the the default exception handler.

Additionally, for the page output on errors, by default we output a JSON object:

https://github.com/ceph/ceph/blob/7b94ee11ac51fc8a0dafa4181b377d8c74bd58ad/src/pybind/mgr/dashboard/module.py#L140

https://github.com/ceph/ceph/blob/7b94ee11ac51fc8a0dafa4181b377d8c74bd58ad/src/pybind/mgr/dashboard/controllers/_helpers.py#L94-L97

If you want to print a custom message, you just have to feed the `message` field in the `cherrypy.HTTPError(status message=...)`, that will be output to the response body.

Pegonzal replied:
Thanks for all the input, I think I addressed it in all the HTTPError's messages.

I still wonder If I'm proving too many "random" messages, or being too verbose on those, as this is not something we usually see around the Dashboard.
But personally I think it's best to provide a clear reason of failure, so I wonder if there should be a global/shared file with error messages for the Dashboard.

Or if there is already any... I did  not see

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -163,13 +163,17 @@ def logout(self):
+            redirect_url = 'auth/oauth2/logout'
         return {
             'redirect_url': redirect_url
         }
```
Setup a `auth_classes.py` to manage all these classes from there

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+
+    @Endpoint(json_response=False, version=None)
+    def login(self):
+        if self._meets_login_requirements():
```
Changed it to explicitly, by configuring the SSO with `ceph dashboard sso enable oauth2`

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -163,13 +163,17 @@ def logout(self):
         if mgr.SSO_DB.protocol == 'saml2':
             return 'auth/saml2/login'
+        elif mgr.SSO_DB.protocol == 'oauth2':
+            return 'auth/oauth2/login'
```
Using encapuslation now

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+
+logger = logging.getLogger('oauth2')
+
+@Router('/auth/oauth2', secure=False)
```
Left it as it is

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+    @Endpoint(json_response=False, version=None)
+    def login(self):
+        if self._meets_login_requirements():
+            access_token = cherrypy.request.headers.get('X-Access-Token')
```
I'll keep it in mind, thanks Ernesto!

Changed it to explicitly set the token

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+
+            if Oauth2JWTManager.is_blocklisted(access_token):
+                logger.error('Failed to login: Invalid access token')
+                raise cherrypy.HTTPError(500, 'Invalid access token')
```
Updated them

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+            url_prefix = prepare_url_prefix(mgr.get_module_option('url_prefix', default=''))
+
+            self._set_token_cookie(url_prefix, access_token)
+            raise cherrypy.HTTPRedirect(f"{url_prefix}/#/login?access_token={access_token}")
```
Updated the code to follow the controller/service pattern

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+            if Oauth2JWTManager.is_blocklisted(access_token):
+                logger.error('Failed to login: Invalid access token')
+                raise cherrypy.HTTPError(500, 'Invalid access token')
+            Oauth2JWTManager.initialize_from_token(access_token)
```
Using now the ` cherrypy.request` object to store the data

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+        token = Oauth2JWTManager.get_token_from_header()
+        self._delete_token_cookie(token)
+
+        idp_url = 'https://192.168.100.100:8081' #get from oauth2-proxy service
```
Getting it from the `iss` field in the provided JWT

Pegonzal commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -153,9 +156,9 @@ def get_username(cls):
+            if 'jti' in dtoken and not cls.is_blocklisted(dtoken['jti']):
                 user = AuthManager.get_user(dtoken['username'])
-                if user.last_update <= dtoken['iat']:
+                if ('iat' in dtoken and user.last_update <= dtoken['iat']) or mgr.SSO_DB.protocol == 'oauth2':
```
Update this, please revisit it

Pegonzal commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
+        try:
+            return cherrypy.request.cookie['token'].value
+        except KeyError:
+            return cherrypy.request.headers.get('X-Access-Token')
```
Completly right, sorry about that. Updated it!

Pegonzal commented on src/pybind/mgr/dashboard/services/auth.py:
```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
+
+    @classmethod
+    def initialize_from_token(cls, token):
+        if not JwtManager.is_blocklisted(token):
```
Yes we are

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+                logger.error('Failed to login: Missing access token')
+                raise cherrypy.HTTPError(400, 'Failed to login: Missing access token')
+
+            if Oauth2JWTManager.is_blocklisted(access_token):
```
Updated

Pegonzal commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,60 @@
+
+        idp_url = 'https://192.168.100.100:8081' #get from oauth2-proxy service
+        response = requests.get(f'{idp_url}/realms/{mgr.SSO_DB.oauth2.realm_name}/protocol/openid-connect/logout?post_logout_redirect_uri={cherrypy.request.base}&id_token_hint={token}')
+        return response
```
following better controller/service pattern now

epuertat commented on src/pybind/mgr/dashboard/controllers/auth_classes.py:
```diff
@@ -0,0 +1,88 @@
+    def __init__(self):
+        pass
+
+    def to_dict(self) -> Dict[str, str]:
+        return {}
+
+    @classmethod
+    def from_dict(cls, s_dict):
+        return super().from_dict(s_dict)
```
Same thing here: if we're expecting some dict with specific fields, then we should declare it.

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,31 @@
+        token = OAuth2.get_token(cherrypy.request)
+
+        raise cherrypy.HTTPRedirect(OAuth2.get_logout_redirect_url(token))
+        # response = requests.get(OAuth2.get_logout_redirect_url(token), verify=False)
+        # return response
```
Remove?

Pegonzal replied:
Oops, yes.

epuertat commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/auth.service.ts:
```diff
@@ -42,12 +42,14 @@ export class AuthService {
       this.authStorageService.remove();
-      const url = _.get(this.route.snapshot.queryParams, 'returnUrl', '/login');
+      if (resp.sso) {
+        return window.location.replace(resp.redirect_url);
```
Looks like [Angular provides a specific method for this](https://angular.dev/api/common/Location). It's always better to use the Angular services, as they might be doing extra stuff (for example, I think that the `window.location.*` might override the Angular routing mechanisms).

Angular still [suggests using `Router.navigate` instead](https://angular.dev/api/common/Location?tab=usage-notes).

Pegonzal replied:
What I'm trying to do here is to redirect to the backend's @Rotuer(/auth/oauth2), tried different angualar router methods, but none did the trick.

The page always ends up in a loop in the /login page, trying to log again. I think this is due to blocklisting the token on the logout process and removing the localStorage user's info so all following operations are not permitted.

So I tried to not altear any of the current implementation and just added the window.location.replace for the oauth2 logout mechanism.

epuertat replied:
Ok!

epuertat commented on src/pybind/mgr/dashboard/module.py:
```diff
@@ -275,6 +275,7 @@ class Module(MgrModule, CherryPyConfig):
                min=400, max=599),
         Option(name='redirect_resolve_ip_addr', type='bool', default=False),
         Option(name='cross_origin_url', type='str', default=''),
+        Option(name='sso_oauth2', type='bool', default=False),
```
Following up with the idea of encapsulating functionally related code, why not:
```suggestion
        from auth import OAuth2
        ...
        MODULE_OPTIONS.extend(OAuth.module_options),
```
At some point in the future, following this pattern we will be creating an interface for new functionality.

Pegonzal replied:
I get the idea and tried to follow it but, for some reason it was breaking, when using the local import like in the suggestion above I was getting:

```
ceph-node-00 ceph-mgr[2737]: mgr[py] Traceback (most recent call last):
                                               File "/usr/share/ceph/mgr/dashboard/__init__.py", line 60, in <module>
                                                 from .module import Module, StandbyModule  # noqa: F401
                                               File "/usr/share/ceph/mgr/dashboard/module.py", line 231, in <module>
                                                 class Module(MgrModule, CherryPyConfig):
                                               File "/usr/share/ceph/mgr/dashboard/module.py", line 279, in Module
                                                 from services.auth.oauth2 import OAuth2
                                             ModuleNotFoundError: No module named 'services'
                                             ```

epuertat commented on src/pybind/mgr/dashboard/services/sso.py:
```diff
@@ -79,20 +66,30 @@ def load(cls):
     mgr.SSO_DB = SsoDB.load()  # type: ignore
 
 
+@CLIWriteCommand("dashboard sso enable oauth2")
+def enable_sso(_):
+    mgr.SSO_DB.protocol = AuthType.OAUTH2
+    mgr.SSO_DB.save()
+    mgr.set_module_option('sso_oauth2', True)
+    return HandleCommandResult(stdout='SSO is "enabled" with "OAuth2" protocol.')
```
No disable command?

What I did, for example, for the debug command is making the enable/disable the parameter of the CLI command (this way you only need 1 command instead of 3, but I'm not sure if you tried to make this compatible with some `ceph dashboard sso enable saml2` command):
```sh
ceph dashboard debug enable
ceph dashboard debug disable
ceph dashboard debug status
```

Pegonzal replied:
There is no disable because it is already implemented as[ `dashboard sso disable` in the saml2 sso cmds], so I just modied that a bit(https://github.com/ceph/ceph/pull/58456/files#diff-51ae22ae2c7aded6a4a8d414641c20b835bba5a810630b55c8ad9dc9eda9451cL150)

epuertat replied:
Ah ok, got it!

epuertat commented on src/pybind/mgr/dashboard/services/sso.py:
```diff
@@ -79,20 +66,30 @@ def load(cls):
-                   Saml2.from_dict(dict_db.get('saml2')))
+        dict_db = json.loads(json_db)  # type: Dict[str, str]
+        protocol: AuthType = AuthType(str(dict_db.get('protocol')))
+        config: SSOAuth = (Saml2.from_dict(str(dict_db.get('saml2'))) if protocol == AuthType.SAML2
+                           else OAuth2.from_dict(str(dict_db.get('oauth2'))))
```
You could also use the `BaseAuth.from_protocol()` routing method?

Pegonzal replied:
Yep, did that

epuertat commented on src/pybind/mgr/dashboard/services/oauth2.py:
```diff
@@ -0,0 +1,132 @@
+            roles = next((value['roles'] for _, value in jwt_payload['realm_access'].items()),
+                         user_roles)
+        else:
+            logger.info('No user roles found, continuing as read-only user')
```
If no role is found, it might be the case that this user shouldn't be allowed any access to Ceph Dashboard. Keep in mind that SSOs in larger companies/users, include all employee accounts, and in many cases only a small subset of them should have access to IT systems (including Ceph).

We can make this more flexible by adding some module option to tune this behaviour (e.g.: `sso_role_fallback` = "deny" or "read-only").

Pegonzal replied:
> We can make this more flexible by adding some module option to tune this behaviour (e.g.: sso_role_fallback = "deny" or "read-only").

That sounds cool, I'll think about that.

> If no role is found, it might be the case that this user shouldn't be allowed any access to Ceph Dashboard.
 
Yeah, I get that. 
My thought process here was that if you've made it this far you have properly configured the `oauth2-proxy` service with your IdP's `client_ID `and `secret`, so only users of that spcified client should be able to get to this part of the code.
Therefore I thought it would be easier for a user that expected to have X permissions to realize that those were not set. Rather than simply not logging in, since well... everything was in place to be logged in!
Also I thought giving read-only access was harmless and therefore I went with it.

I'm not really sure if I should just remove this, as it is true that if you have not properly configured your user, maybe it just shouldn't be able to login.

So your call @epuertat , or anyone who might read all this... Should I remove the default read-only user permissions if no permissions are found in place for the user?

epuertat replied:
@nizamial09 @rkachach @afreen23 @rkachach we summon you!

IMHO this is the issue with authentication vs authorization. All SSO organization users should be valid users to the Ceph Dashboard, but only a few should be authorized. In any case, my proposal would allow for both scenarios, being the "reject" fallback the default based on the [Principle of Least Authority](https://en.wikipedia.org/wiki/Principle_of_least_privilege).

Pegonzal replied:
In the end I went with the `Principle of Least Authority` as you mentioned, and got rid of the default `read-only` role.
Now if no roles are provided, user won't be able to access

rkachach replied:
@Pegonzal @epuertat my vote would be for not letting the user in. "readonly" sounds harmless but it is already giving access to the system and a malicious will have access to see the system configuration and what's going on (maybe preparing for his next attack). IMHO that would be a serious security breach and I'd recommend just refusing the access. If this user really needs to login he will reach out his IT and they will help him to investigate what's going on. My 2 cents.

epuertat commented on src/pybind/mgr/dashboard/services/oauth2.py:
```diff
@@ -0,0 +1,132 @@
+logger = logging.getLogger('OAuth2')
+
+
+class OAuth2:
+    @classmethod
+    def get_token(cls, request: _ThreadLocalProxy) -> str:
```
There's a lot of code in this class and I have the feeling that some amount of it could be shared with local/saml2 implementations? If that's the case, I'd suggest that we have a base class with common methods.

Pegonzal replied:
There are some similarities between this and the local/saml2 as you mentioned.
In the `services/oauth2.py` I can see the following methods that could be inherted for all protocols: `get_token` `set_token`, `get_token_payload`, `set_token_payload`. But even this "common" methods differ a bit from the current local implementation. A part from these, the rest are particular to the oauth2 protocol.

For sure we could have a base class and encapsulate some of the code there and expand on each of the protocols but I think that would required a lot of changes on the current local and saml2 implementation, and I would like to avoid that, for now at least.

epuertat replied:
Ok, that's the benefit of defining an abstract interface: you move the common methods to the ABC class, and then force the implementation in the derived classes. If there's some code that could also be used, you can always inherit the method as well:
```py
class SSOAuth(BaseAuth):
   def get_token(self):
     # logic to parse token from header/cookie
     return token

class OAuth2(SSOAuth):
  def get_token(self):
    token = super().get_token()
    # do extra stuff with the token
```

Pegonzal replied:
[This](https://github.com/ceph/ceph/pull/58456#discussion_r1756205897) follows on with the same thing.

Some refactoring on the JwtManager is neeeded to achive this

epuertat replied:
Ok!

epuertat reviewed (commented):
LGTM! Good work, Pedro! Just a few more things to fix/improve and we're done!

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -156,34 +158,32 @@ def create(self, username, password, ttl: Optional[int] = None):
     def logout(self):
-        logger.debug('Logout successful')
-        token = JwtManager.get_token_from_header()
+        token = JwtManager.get_token(cherrypy.request)
```
You are right, but I don't think it made much sanse to print `Logout successful` when the logout process just started, I'll leave the log and just change the wording

epuertat replied:
You're right! Didn't notice that. In critical functions, you can also log before a function starts and when it finishes (this is better achieved with a decorator, and I'd be surprised if we don't have any for this purpose).

Pegonzal replied:
I couldn't find any decorator for that.
For now I left it as it was and changed the phrase

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth_classes.py:
```diff
@@ -0,0 +1,88 @@
+        return super().from_dict(s_dict)
+
+
+def decode_jwt_segment(encoded_segment: str):
+    # We add ==== as padding to ignore the requirement to have correct padding in
+    # the urlsafe_b64decode method.
+    return json.loads(base64.urlsafe_b64decode(encoded_segment + "===="))
```
I had some issues with circular dependencies, that's why this ended up here...
I'll try to find a better place for it

epuertat replied:
Ah, ok, understood. You can always do `from X import Y` which should avoid many circular deps (unless we have module-scope objects).

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth_classes.py:
```diff
@@ -0,0 +1,88 @@
+from typing import Dict
+
+
+class AuthType(str, Enum):
+    LOCAL = ''
+    SAML2 = 'saml2'
+    OAUTH2 = 'oauth2'
```
It saves me one import, but... It makes you have to write `BaseAuth` every time, and in my opinion it is too verbose and just harder to comprehend at first sight than: `AuthType.OAUTH2`

epuertat commented on src/pybind/mgr/dashboard/controllers/__init__.py:
```diff
@@ -11,6 +11,7 @@
 from ._router import Router
 from ._task import Task
 from ._ui_router import UIRouter
+from .auth_classes import BaseAuth
```
BTW, since the `auth` service is becoming overcrowded, it's ok to split it if smaller files or even a separate dir (I was in fact thinking on having per-backend files in order to support future dynamic loading of auth backends). Something like `services/auth/*` could work, but that would force you to update a few imports over there.

Pegonzal replied:
I went with your suggestion and separated the auth classes in respective files in `services/auth/*`

epuertat commented on src/pybind/mgr/dashboard/controllers/auth_classes.py:
```diff
@@ -0,0 +1,88 @@
+        return self.onelogin_settings['sp']['attributeConsumingService']['requestedAttributes'][0][
+            'name']
+
+    def to_dict(self) -> Dict[str, str]:
+        return {
+            'onelogin_settings': self.onelogin_settings
+        }
+
+    @classmethod
+    def from_dict(cls, s_dict):
+        return Saml2(s_dict['onelogin_settings'])
```
Haha, if you ask an OOP-believer, there's no "too many" when it comes to classes. That said, they usually require some design upfront, so everything is "classy" (in Python everything is an object/class). In this case, we're trying to accomodate code that's 50% procedural 50% OOP.

That said, don't worry since classes can be used as pure/thick OOP classes (data + methods) or plain/think classes (data containers) just for typing purposes. With complex Python projects a this one, you definitely need to be serious about typing, and classes are easier to type than dicts.

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth_classes.py:
```diff
@@ -0,0 +1,88 @@
+    LOGOUT_URL: str
+    sso: bool
+
+    @classmethod
+    def from_protocol(cls, protocol: AuthType) -> "BaseAuth":
+        auth_map = {
+            AuthType.LOCAL: Local,
+            AuthType.SAML2: Saml2,
+            AuthType.OAUTH2: OAuth2
+        }
+        return auth_map[protocol]
```
Subclasses of subclasses are not being taken care of here.
I ended up itereting over the subclass to it's subclasses, i.e. the `SSOAuth` Subclasses: `Saml2` and `OAuth2`

Pegonzal commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -156,34 +158,32 @@ def create(self, username, password, ttl: Optional[int] = None):
     def check(self, token):
         if token:
-            user = JwtManager.get_user(token)
+            if mgr.SSO_DB.protocol == AuthType.OAUTH2:
+                user = OAuth2.get_user(token)
+            else:
+                user = JwtManager.get_user(token)
```
For this some refactor is needed on the JwtManager, I started that work (locally), but too many changes were being added, I left it as is and wrote it down as an improvement TODO in the PR's description

Pegonzal commented:
Thanks @epuertat for all the review efforts here. It is appreciated.

I did update my PR one last time. Also, updated the PR's description with steps to set up the SSO oauth2, updated the video with a full workflow and added a section of TODOs to improve this work.

From your last review there were some suggestions I did end up not following or just wasn't able to. Answered each of them on your review.
I don't think none of them is a blocker, you can confirm that. 
I would like to try to get this merged, and follow up with any other work in new/different PRs.

epuertat commented on src/pybind/mgr/dashboard/controllers/_base_controller.py:
```diff
@@ -8,7 +8,7 @@
 
 from ..plugins import PLUGIN_MANAGER
-from ..services.auth import AuthManager, JwtManager
+from ..services.auth.auth import AuthManager, JwtManager
```
To avoid this ugly redundancy, you usually add the key classes in a directory (subpackage) to the `__all__` variable of the `__init__.py` of that directory. That way you end up importing (which doesn't change the syntax at all):
```suggestion
from ..services.auth import AuthManager, JwtManager
```
This way the new subpackage/module behaves the same as the old one for the rest of the code, but you were able to refactor the into separate files.

Pegonzal replied:
Right, thanks for the tip. Updated it

epuertat commented on src/pybind/mgr/dashboard/controllers/auth.py:
```diff
@@ -10,7 +10,8 @@
 from .. import mgr
 from ..exceptions import InvalidCredentialsError, UserDoesNotExist
-from ..services.auth import AuthManager, JwtManager
+from ..services.auth.auth import AuthManager, AuthType, BaseAuth, JwtManager
+from ..services.auth.oauth2 import OAuth2
```
Same thing here. You could also add `OAuth2` to the `__all__` list, or make it explicit as you did here:
```suggestion
from ..services.auth import AuthManager, AuthType, BaseAuth, JwtManager
from ..services.auth.oauth2 import OAuth2
```

. A good thing of importing through `__init__.py` `__all__` is that you keep the implementation details (e.g.: filenames) hidden:
```suggestion
from ..services.auth import AuthManager, AuthType, BaseAuth, JwtManager, OAuth2
```

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,29 @@
+
+
+@Router('/auth/oauth2', secure=False)
+class Oauth2(BaseController, ControllerAuthMixin):
```
I'm curious, why just `BaseController`? Does it generate the correct OpenAPI doc?
```suggestion
class Oauth2(RESTController, ControllerAuthMixin):
```

Pegonzal replied:
No particular reason other than following the Saml2 implementation and not noticing this.. changed it

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,29 @@
+
+    @Endpoint(json_response=False, version=None)
+    def login(self):
+        if not mgr.get_module_option('sso_oauth2'):
+            raise DashboardException(500, msg='Failed to login: SSO OAuth2 is not enabled')
```
This is still exposing implementation details from the PoV of a controller. The only thing that matters to this controller is whether OAuth2 is enabled or not, but not what you have to do to find that out:
```suggestion
        if not OAuth2.enabled():
            raise DashboardException(500, msg='Failed to login: SSO OAuth2 is not enabled')
```

Pegonzal replied:
Thanks, got it! Updated it

epuertat commented on src/pybind/mgr/dashboard/controllers/oauth2.py:
```diff
@@ -0,0 +1,29 @@
+
+    @Endpoint(json_response=False, version=None)
+    def logout(self):
+        token = OAuth2.get_token(cherrypy.request)
```
For consistency, shouldn't you also check if OAuth2 is enabled here?

Pegonzal replied:
Makes sense, added the check

epuertat reviewed (approved):
LGTM! Just left a few things that I think you should improve before merging, but it's almost 100% mergeable. Great work, Pedro!

rkachach commented:
@Pegonzal I'm getting the following error when cherry-picking your last commit:

```
Sep 13 04:14:23 ceph-node-0 ceph-mgr[2981]: mgr[py] Traceback (most recent call last):
                                              File "/usr/share/ceph/mgr/dashboard/__init__.py", line 60, in <module>
                                                from .module import Module, StandbyModule  # noqa: F401
                                              File "/usr/share/ceph/mgr/dashboard/module.py", line 30, in <module>
                                                from .controllers import Router, json_error_page
                                              File "/usr/share/ceph/mgr/dashboard/controllers/__init__.py", line 1, in <module>
                                                from ._api_router import APIRouter
                                              File "/usr/share/ceph/mgr/dashboard/controllers/_api_router.py", line 1, in <module>
                                                from ._router import Router
                                              File "/usr/share/ceph/mgr/dashboard/controllers/_router.py", line 7, in <module>
                                                from ._base_controller import BaseController
                                              File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 11, in <module>
                                                from ..services.auth import AuthManager, JwtManager
                                            ImportError: cannot import name 'AuthManager' from 'dashboard.services.auth' (unknown location)

```

Pegonzal commented:
> @Pegonzal I'm getting the following error when cherry-picking your last commit:
> 
> 
> 
> ```
> 
> Sep 13 04:14:23 ceph-node-0 ceph-mgr[2981]: mgr[py] Traceback (most recent call last):
> 
>                                               File "/usr/share/ceph/mgr/dashboard/__init__.py", line 60, in <module>
> 
>                                                 from .module import Module, StandbyModule  # noqa: F401
> 
>                                               File "/usr/share/ceph/mgr/dashboard/module.py", line 30, in <module>
> 
>                                                 from .controllers import Router, json_error_page
> 
>                                               File "/usr/share/ceph/mgr/dashboard/controllers/__init__.py", line 1, in <module>
> 
>                                                 from ._api_router import APIRouter
> 
>                                               File "/usr/share/ceph/mgr/dashboard/controllers/_api_router.py", line 1, in <module>
> 
>                                                 from ._router import Router
> 
>                                               File "/usr/share/ceph/mgr/dashboard/controllers/_router.py", line 7, in <module>
> 
>                                                 from ._base_controller import BaseController
> 
>                                               File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 11, in <module>
> 
>                                                 from ..services.auth import AuthManager, JwtManager
> 
>                                             ImportError: cannot import name 'AuthManager' from 'dashboard.services.auth' (unknown location)
> 
> 
> 
> ```

I literally just updated the PR because of that.
Could you try again @rkachach?

rkachach commented:
> > @Pegonzal I'm getting the following error when cherry-picking your last commit:
> > ```
> > 
> > Sep 13 04:14:23 ceph-node-0 ceph-mgr[2981]: mgr[py] Traceback (most recent call last):
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/__init__.py", line 60, in <module>
> > 
> >                                                 from .module import Module, StandbyModule  # noqa: F401
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/module.py", line 30, in <module>
> > 
> >                                                 from .controllers import Router, json_error_page
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/controllers/__init__.py", line 1, in <module>
> > 
> >                                                 from ._api_router import APIRouter
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/controllers/_api_router.py", line 1, in <module>
> > 
> >                                                 from ._router import Router
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/controllers/_router.py", line 7, in <module>
> > 
> >                                                 from ._base_controller import BaseController
> > 
> >                                               File "/usr/share/ceph/mgr/dashboard/controllers/_base_controller.py", line 11, in <module>
> > 
> >                                                 from ..services.auth import AuthManager, JwtManager
> > 
> >                                             ImportError: cannot import name 'AuthManager' from 'dashboard.services.auth' (unknown location)
> > ```
> 
> I literally just updated the PR because of that. Could you try again @rkachach?

Still failing ... did you miss to submit the `__init__.py` file which Ernesto has proposed to add in the auth directory?

Pegonzal commented:
jenkins test windows

Pegonzal commented:
API tests OK
make check failed with:

```
dashboard/services/auth/__init__.py:1: note: In module imported here,
dashboard/module.py:33: note: ... from here,
dashboard/__init__.py:60: note: ... from here:
dashboard/services/auth/auth.py: note: In member "to_dict" of class "Local":
dashboard/services/auth/auth.py:89: error: Incompatible return value type (got "Config", expected "Config")  [return-value]
```

Fixed it with 

```
     def to_dict(self) -> 'BaseAuth.Config':
-        return self.Config()
+        return BaseAuth.Config()
```

Failure solved locally:
```
mypy: commands[0]> mypy --config-file=../../mypy.ini -m alerts -m balancer -m cephadm -m crash -m dashboard -m devicehealth -m diskprediction_local -m hello -m influx -m iostat -m localpool -m mds_autoscaler -m mgr_module -m mgr_util -m mirroring -m nfs -m orchestrator -m pg_autoscaler -m progress -m prometheus -m rbd_support -m rgw -m rook -m selftest -m smb -m snap_schedule -m stats -m status -m telegraf -m telemetry -m test_orchestrator -m volumes -m zabbix
Success: no issues found in 33 source files
  mypy: OK (3.49=setup[0.07]+cmd[3.42] seconds)
  congratulations :) (3.60 seconds)
```

Also added a fix for previously identified bug, when loading the SSO_DB:

```
Traceback (most recent call last):
                                              File "/usr/share/ceph/mgr/dashboard/module.py", line 336, in serve
                                                load_sso_db()
                                              File "/usr/share/ceph/mgr/dashboard/services/sso.py", line 93, in load_sso_db
                                                mgr.SSO_DB = SsoDB.load()  # type: ignore
                                              File "/usr/share/ceph/mgr/dashboard/services/sso.py", line 89, in load
                                                Saml2.from_dict(dict_db.get('saml2')))
                                              File "/usr/share/ceph/mgr/dashboard/services/sso.py", line 42, in from_dict
                                                return Saml2(s_dict['onelogin_settings'])
                                            KeyError: 'onelogin_settings'
  ```
  
  Fix: 
  
  ```
    @classmethod
     def from_dict(cls, s_dict):
-        return Saml2(s_dict['onelogin_settings'])
+        try:
+            return Saml2(s_dict['onelogin_settings'])
+        except KeyError:
+            return Saml2({})
  ```

Pegonzal commented:
Found bug, the SSO_DB was not backwards-compatible due to local protocol changing from: `''` to `AuthType.LOCAL`

Error found:
```
                                              File "/usr/share/ceph/mgr/dashboard/module.py", line 337, in serve
                                                load_sso_db()
                                              File "/usr/share/ceph/mgr/dashboard/services/sso.py", line 83, in load_sso_db
                                                mgr.SSO_DB = SsoDB.load()  # type: ignore
                                              File "/usr/share/ceph/mgr/dashboard/services/sso.py", line 77, in load
                                                protocol: AuthType = AuthType(str(dict_db.get('protocol')))
                                              File "/lib64/python3.9/enum.py", line 384, in __call__
                                                return cls.__new__(cls, value)
                                              File "/lib64/python3.9/enum.py", line 702, in __new__
                                                raise ve_exc
                                            ValueError: '' is not a valid AuthType
  ```

Fixed it with:

```
-        protocol: AuthType = AuthType(str(dict_db.get('protocol')))
-        config: BaseAuth = BaseAuth.from_protocol(protocol).from_dict(str(dict_db.get(protocol)))
+        protocol = dict_db.get('protocol')
+        # keep backward-compatibility
+        if protocol == '':
+            protocol = AuthType.LOCAL
+        protocol = AuthType(protocol)
+        config: BaseAuth = BaseAuth.from_protocol(protocol).from_dict(dict_db.get(protocol))

```

rkachach commented:
Using the code from the last commit, I can create the service `oauth2-proxy` without issues:

```
[ceph: root@ceph-node-0 /]# ceph orch ls --service_name oauth2-proxy --export
service_type: oauth2-proxy
service_name: oauth2-proxy
placement:
  count: 1
  hosts:
  - ceph-node-2
spec:
  allowlist_domains:
  - domain1.com
  - 192.168.100.1:8080
  - 192.168.100.101
  - ceph-node-1
  client_id: oauth2-proxy
  client_secret: vwCLTITb2rtem583523uXcuSNItDsz1I
  oidc_issuer_url: http://192.168.100.1:8080/realms/ceph
  provider_display_name: Redo My OIDC provider
```

rkachach reviewed (approved):
LGTM (I tested the code and it works without issues)

epuertat commented on src/pybind/mgr/dashboard/services/auth/saml2.py:
```diff
@@ -0,0 +1,35 @@
+
+    @classmethod
+    def from_dict(cls, s_dict: Saml2Config) -> 'Saml2':
+        try:
+            return Saml2(s_dict['onelogin_settings'])
+        except KeyError:
+            return Saml2({})
```
One very subtle comment here. Imagine that you didn't write `Saml()` code and don't know whether it also triggers `KeyError`... In general it's good that try-except blocks catch only what's missing:
```suggestion
        try:
            config = s_dict['onelogin_settings'])
        except KeyError:
            config = {}
        finally:
            return Saml2(config)
```

An alternative approach (less explicit):
```suggestion
        config = s_dict.get('onelogin_settings', {})
        return Saml2(config)
```

Pegonzal replied:
Makes sense. Updated it. Thanks!

epuertat commented on src/pybind/mgr/dashboard/services/auth/auth.py:
```diff
@@ -25,6 +30,74 @@
+    def from_dict(cls, s_dict) -> 'BaseAuth':
+        pass
+
+    @classmethod
+    def get_auth_name(cls):
+        return cls.__name__.lower()
```
Why is this defined in every subclass?
```py
>>> class A:
...   @classmethod
...   def name(cls):
...     return cls.__name__.lower()
... 
>>> A.name()
'a'
>>> class B(A): pass
... 
>>> B.name()
'b'
```

Pegonzal replied:
OOP noob I guess...
First time I tried it I was getting an error ( don't remember which one), so I ended up with that.
Something must have been messed up in the inhertance before.

Anyways, I have updated it to remove the repetition

epuertat commented on src/pybind/mgr/dashboard/services/auth/auth.py:
```diff
@@ -25,6 +30,74 @@
+    def to_dict(self) -> 'BaseAuth.Config':
+        return BaseAuth.Config()
+
+    @classmethod
+    def from_dict(cls, s_dict: BaseAuth.Config) -> 'Local':
+        # pylint: disable=unused-argument
+        return cls()
```
If an abstract method in a derived class doesn't make sense (not ideal as it breaks the Liskov substitution principle), feel free to act accordingly:
```suggestion
    @classmethod
    def from_dict(cls, s_dict: BaseAuth.Config) -> 'Local':
        raise NotImplementedError
```

Pegonzal replied:
I guess this might be a design problem more than a implemntation one as it doesn't make sense to have this methods for the `Local` and `Oauth2` but it does for `Saml2`

And it can not raise a `NotImplementedError` because on `load_sso_db`  the `from_dict` method of the active protocol will be called. [This is what I mean](https://github.com/ceph/ceph/blob/c37a00af2113d089d40b9d6df347fcb940efa41a/src/pybind/mgr/dashboard/services/sso.py#L82)
 
 What do you think @epuertat ?

epuertat replied:
It's ok then!

epuertat commented on src/pybind/mgr/dashboard/services/auth/oauth2.py:
```diff
@@ -0,0 +1,151 @@
+    def to_dict(self) -> 'BaseAuth.Config':
+        return self.OAuth2Config()
+
+    @classmethod
+    def from_dict(cls, s_dict: OAuth2Config) -> 'OAuth2':
+        # pylint: disable=unused-argument
+        return OAuth2()
```
Same thing here: what's the sense of this? It's ignoring the input `s_dict`.

Pegonzal replied:
Same thing as before, and it's reciving the `s_dict` because of the base class implementation (should I remove the argument from there?) and it's ignoring the `s_dict` because the OAuth2 does not require any config.

I left it as it is to keep consistency with the Saml2 and with the idea to, maybe in the future and if needed, add config parameters to the OAuth2 and unify the Saml2 and OAuth2 implementations

Pegonzal commented on src/pybind/mgr/dashboard/services/auth/auth.py:
```diff
@@ -25,6 +30,74 @@
+    LOGOUT_URL: str
+    sso: bool
+
+    @staticmethod
+    def from_protocol(protocol: AuthType) -> Type["BaseAuth"]:
+        for subclass in BaseAuth.__subclasses__():
```
Did you change this? I did not, it was a @staticmethod in your example and I followed it blindly withoutnoticing that absurdity.

Updated it.

epuertat replied:
Ah, I copied that from my PR here https://github.com/ceph/ceph/pull/59509/files#diff-17c3c754462ec3d2e3614f52001a501bed033a28fc0d7111c0944817b8446ffbR284-R290, and there this method was under another class, not the ABC.

Pegonzal commented on src/pybind/mgr/dashboard/services/auth/auth.py:
```diff
@@ -25,6 +30,74 @@
+    def get_auth_name(cls):
+        return cls.__name__.lower()
+
+    def to_dict(self) -> 'BaseAuth.Config':
+        return BaseAuth.Config()
```
> This is not using self, hence it's not an instance method but a classmethod. It's returning a default empty object.

My bad, I was previously using self but I was hitting a typeError with the return type, and after changing it I missed this.
It's not using the class so it would be a `staticmethod` then, right?

However, having said that, the reason why I wanted `to_dict` to be an instance method was to access the class instance values as was [being done in the Saml2](https://github.com/ceph/ceph/blob/c37a00af2113d089d40b9d6df347fcb940efa41a/src/pybind/mgr/dashboard/services/auth/saml2.py#L14).

If I change this to a `staticmethod` that would mean I need to change the way of storting the auth config values, so instead of doing it at the object level, should I do it at the class level?

epuertat replied:
That's my point. A `to_dict()` method is expected to "serialize" the properties of a class **instance** as a dict, but in this case it does nothing with `self` and instead returns a generic dict from another base class.

Pegonzal commented on src/pybind/mgr/dashboard/services/auth/auth.py:
```diff
@@ -25,6 +30,74 @@
+        return cls()
+
+
+class SSOAuth(BaseAuth):
+    sso = True
```
Got it, thanks!

Pegonzal commented:
Thaks @epuertat for your last review and sorry I missed it before merging it...
I've opened this PR: #59867 following your latest comments
